Allow multiple references to the same footnote

jsma opened the pull request:
This is an attempt to fix #25

At present, this only ensures that each linked reference will have a unique HTML `id` so that the back links on the footnotes will work (by linking to the first reference). This does not yet include UI changes to expose links to each individual reference in the content but there is a proposed approach in https://github.com/torchbox/wagtail-footnotes/issues/25#issue-1149764004.

zerolab commented:
@jsma I know this is quite late, but would you be able to add a test or two for this?
Also, can you rebase on `main`. We switched to ruff for a bunch of linting and I suspect that may be introducing a conflict

jsma commented:
I completely reworked my initial implementation which required some force pushes. 

I don't know what planet I was on when I first did this but I fixed the `<sup>` to be inside the `<a>` rather than vice versa. This does introduce changes to the `footnotes.html` template, since that was the only way I could test that unique `id` attributes were generated (I opted to use plain 1-based index integers for each reference per [this comment](https://github.com/torchbox/wagtail-footnotes/issues/25#issuecomment-1283125223)). I realize folks may be opinionated on this but at the end of the day, it's a template that can be overridden. 

Sorry for any previous CI noise. I added Python 3.11 type hints out of habit, which didn't fail when I ran `tox` locally against all environments, I guess because it was running in a 3.11-based virtual environment (I'm not super familiar with tox, I maintain websites vs reusable packages so I'm only family with standard Django tests).

jsma commented on tests/test/test_blocks.py:
```diff
@@ -38,19 +38,47 @@ def setUp(self):
                         "type": "paragraph",
-                        "value": f'<p>This is a paragraph with a footnote. <footnote id="{uuid}">1</footnote></p>',
+                        "value": (
+                            f'<p>This is a paragraph with a footnote. <footnote id="{uuid}">[{uuid[:6]}]</footnote></p>'
```
The `[{uuid[:6]}]` bit isn't strictly required since `FIND_FOOTNOTE_TAG` doesn't care what's between `<footnote></footnote>` but I opted to make it an accurate representation of what is actually stored in the database.

zerolab commented on wagtail_footnotes/blocks.py:
```diff
@@ -37,17 +40,28 @@ def replace_footnote_tags(self, value, html, context=None):
         page = new_context["page"]
         if not hasattr(page, "footnotes_list"):
             page.footnotes_list = []
```
I think we can do away with `footnotes_list` in favour of the new `footnotes_references`.

jsma replied:
These are two different things. `footnotes_list` is the list of the actual Footnote objects to be rendered at the bottom of the rendered page, and `footnotes_references` is a dict that maps a Footnote's UUID to a list of all of the references found in the rich text blocks, so that jump links back to each individual reference can be rendered. 

I didn't have any better idea of how to make the Footnote objects aware of the one or more references contained in the rich text blocks. This is why I made the goofy template tag so I could at least do dict lookups using the UUID in the template to fetch the references.

I'm working on getting this rebased on main now that the custom template PR was merged.

zerolab replied:
Right, that makes sense. I think some additional comments would be helpful then, and perhaps some type hinting, then there will be no ambiguity. Thank you

zerolab commented on wagtail_footnotes/templatetags/wagtailfootnotes_tags.py:
```diff
@@ -0,0 +1,15 @@
+    """
+    if hasattr(value, "footnotes_references"):
+        return value.footnotes_references.get(footnote_uuid, [])
+    return []
```
I am no sure we need this. if we drop `footnotes_list`, we can update the template to check for `page.footnotes_references` and tweak the loops accordingly

zerolab commented on wagtail_footnotes/blocks.py:
```diff
@@ -37,17 +40,28 @@ def replace_footnote_tags(self, value, html, context=None):
+                # 0-based) and if it's the second link to the first footnote, it will be "footnote-source-1-1", etc.
+                # This ensures the ids are unique throughout the page and allows for the template to generate links from
+                # the footnote back up to the distinct references in the content.
+                link_id = f"footnote-source-{index}-{len(page.footnotes_references[footnote_uuid])}"
```
suggestion: keep indexes consistent. for regular users 1-1, 1-2 etc make more sense than 1-0. 1-1

zerolab reviewed (changes requested):
It took quite some time to get to review this. Sorry about that.
Left a few thoughts and suggestions for consistency and some simplification

zerolab commented:
uff, this clashes a bit with #70
Could you rebase again @jsma then I can do a final test, re-review and merge in.
Sorry it took so long

jsma committed:
Allow multiple references to the same footnote

Previously every link in the rich text for the same footnote had the same `id` attribute, which breaks the "Back to content" link (or just always returns the user to the first reference).
This now generates a unique `id` for each footnote reference and updates the `footnotes.html` template to generate unique links back to each reference in the content.

jsma commented on wagtail_footnotes/blocks.py:
```diff
@@ -15,69 +16,93 @@ class RichTextBlockWithFootnotes(RichTextBlock):
         html = super().render(value, context=context)
         return self.replace_footnote_tags(value, html, context=context)
 
-    def render_basic(self, value, context=None):
```
I removed `render_basic` because `super.render()` will already return this, and at the end of the day, all that is happening is regex substitutions on the output, which is handled in `render()` above.

jsma commented on wagtail_footnotes/blocks.py:
```diff
@@ -15,69 +16,93 @@ class RichTextBlockWithFootnotes(RichTextBlock):
+            footnote_uuid = match.group(1)
             try:
-                index = self.process_footnote(match.group(1), page)
-            except (KeyError, ValidationError):
```
I don't see anything that would raise a `ValidationError`, so I removed it.

jsma commented:
I've finally rebased this on main and incorporated the template-based rendering changes introduced in #70.

I reworked the implementation a bit to avoid the goofy template tag I relied on previously. It now patches the footnotes themselves before appending them to the page, so the footnote will have knowledge of how many times it has been referenced in the text above. I didn't have any bright ideas beyond just tracking 1-indexed integers so that at least footnote_reference.html and footnotes.html would agree on what the `id` on the reference is, and it would parallel how we're using an index to set the `id` for the footnote.

kevinhowbrook commented on wagtail_footnotes/templates/wagtail_footnotes/includes/footnotes.html:
```diff
@@ -7,10 +7,27 @@ <h2 id="footnote-label">
+                <li id="footnote-{{ footnote_index }}">
+                    {% if footnote.references|length == 1 %}
+                        {# If there is only a single reference, link the return icon back to it #}
+                        <a href="#footnote-source-{{ footnote_index }}-{{ footnote.references.0 }}" aria-label={% translate "Back to content" %}>
```
**issue**
Missing quotes for aria-label

```suggestion
                        <a href="#footnote-source-{{ footnote_index }}-{{ footnote.references.0 }}" aria-label="{% translate "Back to content" %}">
```

kevinhowbrook commented on wagtail_footnotes/templates/wagtail_footnotes/includes/footnotes.html:
```diff
@@ -7,10 +7,27 @@ <h2 id="footnote-label">
+                        ↩
+                        {% for reference_index in footnote.references %}
+                            {# If there are multiple references, generate unique links per reference #}
+                            <a href="#footnote-source-{{ footnote_index }}-{{ reference_index }}" aria-label={% translate "Back to content" %}>
```
**Issue**
Missing quotes for aria-label

```suggestion
                            <a href="#footnote-source-{{ footnote_index }}-{{ reference_index }}" aria-label="{% translate "Back to content" %}">
```

kevinhowbrook commented on tests/test/test_functional.py:
**Potential follow-up PR / New issue to raise from this** 

Looking at the tests, specifically test_functional.py we don't have tests for multiple references. There's no test that renders the page and checks the footnotes list HTML contains the multiple back-links. The block-level test covers the in-text anchors but nothing verifies the footnotes.html output for this case end-to-end.

All tests use a single UUID. We don't have a test for a page with footnote A and footnote B each referenced once, verifying they get footnote-source-1-1 and footnote-source-2-1, and that the ordering in footnotes_list is correct.

Same footnote referenced across multiple blocks
The multiple-reference test puts both references inside the same block value. A more realistic case is two separate paragraph blocks each referencing the same footnote. Since page.footnotes_list persists across block renders, this is a meaningful edge case that could behave differently.

Mixed case/multiple footnotes, one with multiple references
A page with footnote A referenced twice and footnote B referenced once, verifying the reference counters don't bleed across different footnotes (i.e. footnote B still gets footnote-source-2-1, not footnote-source-2-3).

(I'm happy to pick this up if we don't want to address it here)

kevinhowbrook reviewed (commented):
Thanks so much for your continued work on this @jsma, here's a review of the code from me where it stands today. 

**Testing notes:**
Preview works well
<img width="1603" height="623" alt="Image" src="https://github.com/user-attachments/assets/3f6b2bc6-adf1-4d3c-bf5d-f2c432421747" />

Rendered page looks good
<img width="1836" height="640" alt="Image" src="https://github.com/user-attachments/assets/066803c8-7b19-4069-bb71-f48749161f04" />

I'd suggest we update the CHANGELOG and bump the version too in https://github.com/torchbox/wagtail-footnotes/blob/main/wagtail_footnotes/__init__.py to VERSION = (0, 15, 0) because of:

**Possible breaking change to HTML IDs no changelog**
Every existing site has id="footnote-source-1" in their HTML, CSS, JS, tests, etc. This PR changes all IDs to footnote-source-1-1 unconditionally, even single references. We should probably mention that in the CHANGELOG.

**[N] prefix silently dropped**
The original template rendered [1] text . The new template removes the [N] prefix and looks to me like numbering is now only implied by the `<ol>` counter. Thats a visible change for existing users, I doubt anyone is binding/targeting that for anything like CSS, but I think it's worth pointing out in the change log.

kevinhowbrook commented:
I've changed the target for this PR so I can merge it and pick it up. Thanks @jsma 

Carried on with https://github.com/torchbox/wagtail-footnotes/pull/90
